Add infra/ using Terraform.  Remove vdc/. (#9713)
hail-is/hail: infra/README.md

This is a work in progress to use Terraform to manage our cloud
infrastructure.

Instructions:

- You will need a GCP project.  We assume `gcloud` is configured to
  point at your project.

> 

Reviewer: ```suggestion
gcloud config set project <project-id>
gcloud config set compute/zone <compute-zone>
```

- Create a service account for Terraform with Owner role, create a
  service account key and place it in
  `$HOME/.hail/terraform_sa_key.json`.

- Enable the the GCP services needed by Hail:

   ```
   gcloud services enable \
       compute.googleapis.com \
       cloudresourcemanager.googleapis.com \
       servicenetworking.googleapis.com \
       sqladmin.googleapis.com \
       container.googleapis.com \
       serviceusage.googleapis.com
   ```

- Install terraform.

- Create `$HOME/.hail/global.tfvars` that looks like:

   ```
   gcp_project = "<gcp-project>"
   gcp_region = "<gcp-region>"
   gcp_zone = "<gcp-zone>"
   domain = "<domain>"
   ```

- Run `terraform init`.

- Run `terraform apply -var-file="$HOME/.hail/global.tfvars"`.
  Terraform has created a GKE cluster named `vdc`.  We assume
  `kubectl` is configured to point at this cluster.

> 

Reviewer: ```suggestion
gcloud container clusters get-credentials vdc
```

> - Go to the Google Cloud console, VPC networks > internal > Private

Reviewer: I think you might be able to express this in Terraform using `google_compute_network_peering` (https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_network_peering#export_custom_routes), but I haven't looked into it in detail.

Author: Yeah, that might work.  I tried it but couldn't get it to work (probably user error).  The docs are written to peer two user-created networks, but we want to peer our internal network with the Google managed Cloud SQL network.  I imagine we'll want to make a pass over all of this to make it more automated.  Let's revisit this then.

  service connection > Private connections to services, and enable
  Export custom routes to both connections.

You can now install Hail:

> - Run `$HAIL/docker/third-party/copy_images.sh`.  This copies some

Reviewer: This step also requires setting the `PROJECT` environment variable beforehand and running `gcloud auth configure-docker gcr.io`.

I felt this was slightly faster running this from a VM, so maybe it would be good to combine this step with the VM setup further below.

  base images from Dockerhub (which now has rate limits) to GCR.

- Generate TLS certificates.  See ../dev-docs/tls-cookbook.md.

- Update $HAIL/config.mk with your infrastructure settings.  You can
  get settings from the default/global-config secret.

> 

Reviewer: ```suggestion
kubectl get secret global-config -o json | jq '.data | map_values(@base64d)'
```

> - Run `kubectl -n default apply -f $HAIL/ci/bootstrap.yaml`.

Reviewer: Currently, `KUBERNETES_SERVER_URL` is hardcoded in ci/bootstrap.py, so probably should also be updated?

Author: I know grab it from the env.  Thanks!

> - Build the CI utils image.  Run `make push-ci-utils` in $HAIL/ci.

Reviewer: This failed without running `make service-base` in $HAIL/docker first.

Author: > This failed without running make service-base in $HAIL/docker first.

There should be dependencies for this, I don't see  what's wrong.  push-ci-utils depends on build-ci-utils, and:

```
build-ci-utils:
	$(MAKE) -C ../docker build
	...
```

and in docker/Makefile:

```
build: base-stmp service-base hail-ubuntu
```

> When running this on a Ubuntu VM, I also encountered an odd issue in that Makefile, when stat is called: [ "$(shell bash stat-permissions.sh Dockerfile.base)" = "644" ]. The mode of those files was 664 for me, so I had to run a chmod g-w to get the build working.

Yeah, this is a check for docker caching.  Docker includes files permissions in the layer hash, so unless the permissions agree everywhere, you get painful docker rebuilds.  I added a not about seeing the umask on the VM.  I added a step to set the umask on the VM.

This is a work in progress to use Terraform to manage our cloud
infrastructure.

Instructions:

- You will need a GCP project.  Configure `gcloud` to point at your project:

   ```
   gcloud config set project <gcp-project-id>
   gcloud config set compute/zone <gcp-zone>
   ```

- Create a service account for Terraform with Owner role, create a
  service account key and place it in
  `$HOME/.hail/terraform_sa_key.json`.

- Enable the the GCP services needed by Hail:

   ```
   gcloud services enable \
       compute.googleapis.com \
       cloudresourcemanager.googleapis.com \
       servicenetworking.googleapis.com \
       sqladmin.googleapis.com \
       container.googleapis.com \

>        serviceusage.googleapis.com

Reviewer: cloudprofiler.googleapis.com might be good, too.

   ```

- Install terraform.

- Create `$HOME/.hail/global.tfvars` that looks like:

   ```
   gcp_project = "<gcp-project-id>"
   gcp_region = "<gcp-region>"
   gcp_zone = "<gcp-zone>"
   domain = "<domain>"
   ```

- Run `terraform init`.

- Run `terraform apply -var-file="$HOME/.hail/global.tfvars"`.
  Terraform has created a GKE cluster named `vdc`.  Configure `kubectl` to point at the vdc cluster:

  ```
  gcloud container clusters get-credentials vdc
  ```

- Go to the Google Cloud console, VPC networks > internal > Private
  service connection > Private connections to services, and enable
  Export custom routes to both connections.

You can now install Hail:

- Create a VM on the internal network, 100GB, Ubuntu 20.04 TLS, allow
  full access to all Cloud APIs.  10GB will run out of space.  We

>   assume the rest of the commands are run on the VM.

Reviewer: This pretty much worked for everything except the $HAIL/letsencrypt step, due to a permissions error. I ended up running that outside the VM, but it probably wouldn't be too difficult to fix:

```
Error from server (Forbidden): error when creating "service.yaml": roles.rbac.authorization.k8s.io is forbidden: User "<redacted>" cannot create resource "roles" in API group "rbac.authorization.k8s.io" in the namespace "default": requires one of ["container.roles.create"] permission(s).
Error from server (Forbidden): error when creating "service.yaml": rolebindings.rbac.authorization.k8s.io is forbidden: User "<redacted>" cannot create resource "rolebindings" in API group "rbac.authorization.k8s.io" in the namespace "default": requires one of ["container.roleBindings.create"] permission(s).
make: *** [Makefile:30: run] Error 1
```

Reviewer: This works when using e.g. the terraform service account for running the VM.

Also it seems good to use at least 32 GB of RAM for the VM (otherwise compiling the query service didn't work for me).

Author: Great, I now recommend a n2-standard-8, and to run the VM as a the terraform service account.

- Standardize file permissions.  This is for docker, which considers
  permissions for caching.  Run `echo 'umask 022' > ~/.profile`.  You
  will need to log out/in or run `umask 022`.

- Update $HAIL/config.mk with your infrastructure settings.  You can
  get settings from the default/global-config secret:

  ```
  kubectl get secret global-config -o json | jq '.data | map_values(@base64d)'
  ```

- Install some dependencies on the VM:

  ```
  sudo apt update
  sudo apt install -y docker.io python3-pip
  sudo usermod -a -G docker $USER
  gcloud -q auth configure-docker

>   gcloud container clusters get-credentials --zone us-central1-a vdc

Reviewer: ```suggestion
  gcloud container clusters get-credentials --zone <gcp-zone> vdc
```

  git clone https://github.com/cseed/hail.git
  python3 -m pip install -r $HOME/hail/docker/requirements.txt
  ```

  You will have to log out/in for usermod to take effect.

> - Run `PROJECT=<gc-project-id>

Reviewer: ```suggestion
- Run `PROJECT=<gcp-project-id>
```

  $HAIL/docker/third-party/copy_images.sh`.  This copies some base
  images from Dockerhub (which now has rate limits) to GCR.

- Generate TLS certificates.  See ../dev-docs/tls-cookbook.md.

- Run `kubectl -n default apply -f $HAIL/ci/bootstrap.yaml`.

- Build the CI utils image.  Run `make push-ci-utils` in $HAIL/ci.

- Deploy the bootstrap gateway.  Run `make deploy` in

>   $HAIL/bootstrap-gateway.

Reviewer: This gateway backends will crash-loop, as the router service hasn't been deployed yet:
```
nginx: [emerg] host not found in upstream "router" in /etc/nginx/conf.d/gateway.conf:73
```

That's not a big deal, as the bootstrapping later will take care of it, but maybe worth mentioning.

> - Create Let's Encrypt certs. Run `make run` in $HAIL/letsencrypt.

Reviewer: This step required increasing `server_names_hash_bucket_size` for letsencrypt's nginx, similar to the gateway.

> - Deploy the gateway.  Run `make deploy` in $HAIL/gateway.

Reviewer: This gateway backends will crash-loop, as the router service hasn't been deployed yet:
```
nginx: [emerg] host not found in upstream "router" in /etc/nginx/conf.d/gateway.conf:73
```
That's not a big deal, as the bootstrapping later will take care of it, but maybe worth mentioning.

Author: Thanks!  Yeah, we've run into this kind of issue before.  The k8s services for the Hail services are all in the router deployment, which also runs nginx.  I moved the router service from the router deployment to the gateway deployment.

- Deploy the internal-gateway.  Run `make deploy` in $HAIL/internal-gateway.

- Go to the Google Cloud console, API & Services, Credentials.

>   Configure the consent screen.  Add the scope:

Reviewer: ```suggestion
  Configure the consent screen for internal users.  Add the scope:
```

>   https://www.googleapis.com/auth/userinfo.email.  Create an OAuth

Reviewer: ```suggestion
  https://www.googleapis.com/auth/userinfo.email.  Back in Credentials, create an OAuth
```

  client ID.  Authorize the redirect URI:

>   https://auth.<domain>/oauth2callback.  Download the client secret

Reviewer: For `hailctl auth login` to work, http://127.0.0.1/oauth2callback needs to be added as a redirect URI too.

  as client_secret.json.  Create the auth-oauth2-client-secret secret
  with:

  ```
  kubectl -n default create secret generic auth-oauth2-client-secret --from-file=./client_secret.json
  ```

- Bootstrap the cluster by running:

  ```

>   HAIL_SHA=$(git rev-parse HEAD) \
>   HAIL_CI_UTILS_IMAGE=gcr.io/<gcp-project>/ci-utils:latest \
>   HAIL_CI_BUCKET_NAME=dummy \
>   KUBERNETES_SERVER_URL='<k8s-server-url>' \
>   HAIL_DEFAULT_NAMESPACE='default' \
>   HAIL_DOMAIN=<domain> \
>   HAIL_GCP_ZONE=<gcp-zone> \
>   GCP_PROJECT=<gcp-project> \
>   PYTHONPATH=$HOME/hail/ci:$HOME/hail/batch:$HOME/hail/hail/python \
>   python3 $HAIL/ci/bootstrap.py

Reviewer: ```suggestion
  cd $HOME/hail
  export HAIL_SHA=$(git rev-parse HEAD) \
  export HAIL_CI_UTILS_IMAGE=gcr.io/<gcp-project-id>/ci-utils:latest \
  export HAIL_CI_BUCKET_NAME=dummy \
  export KUBERNETES_SERVER_URL='<k8s-server-url>' \
  export HAIL_DEFAULT_NAMESPACE='default' \
  export HAIL_DOMAIN=<domain> \
  export HAIL_GCP_ZONE=<gcp-zone> \
  export HAIL_GCP_PROJECT=<gcp-project-id> \
  export PYTHONPATH=$HOME/hail/ci:$HOME/hail/batch:$HOME/hail/hail/python \
  python3 ci/bootstrap.py
```

>   python3 $HAIL/ci/bootstrap.py

Reviewer: In order for create_accounts/main to succeed, I had to add a "ci-gsa-key" secret:
https://github.com/populationgenomics/hail/commit/46bbd7e29e7cf092a2d7566b9592df1b63690157

There's still an error at the very end of bootstrap.py, but it's probably just a harmless race?
```
ERROR:asyncio:Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7f8895f01340>
ERROR:asyncio:Unclosed connector
connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x7f8895ecefa0>, 3555.901845417)]']
connector: <aiohttp.connector.TCPConnector object at 0x7f8895f016d0>
```

>   ```

Reviewer: The gateway kept redirecting to hail.is, as it's hardcoded in hail/python/hailtop/config/deploy_config.py (and a few other places):
https://github.com/populationgenomics/hail/commit/b1c3c3fa59b37bcffb0820e50008cd5c270c323f

> - You may want to add a suitable ssh forward rule to the internal network.

Reviewer: This should probably precede the previous point, as otherwise one won't be able to ssh into the VM to run the bootstrap command.

How about this?

Add an "allow-ssh" network tag to your VM and then create the following firewall rule to allow connections to your instance in the internal network:
```
gcloud compute firewall-rules create internal-allow-ssh --network internal --action=allow --target-tags=allow-ssh --rules=tcp:22
```

Author: OK, I added advice about ssh to the instructions for spinning up the VM.  We have custom firewall rules for ssh that aren't going to generalize and I think different users will have different security postures, so OK with this being hand managed for now.

Instructions:

- You will need a GCP project.  Configure `gcloud` to point at your project:

   ```
   gcloud config set project <gcp-project-id>
   gcloud config set compute/zone <gcp-zone>
   ```

- Delete the default network if it exists.

- Create a service account for Terraform with Owner role, create a
  service account key and place it in
  `$HOME/.hail/terraform_sa_key.json`.

- Enable the the GCP services needed by Hail:

   ```
   gcloud services enable \
       compute.googleapis.com \
       cloudresourcemanager.googleapis.com \
       servicenetworking.googleapis.com \
       sqladmin.googleapis.com \
       container.googleapis.com \
       serviceusage.googleapis.com \
       dns.googleapis.com \
       logging.googleapis.com \
       cloudprofiler.googleapis.com
   ```

- Install terraform.

- Create `$HOME/.hail/global.tfvars` that looks like:

   ```
   gsuite_organization = "<gsuite-organization>"
   # batch_gcp_regions is a JSON array of string, the names of the gcp
   # regions to schedule over in Batch.
   batch_gcp_regions = "<batch-gcp-regions>"
   gcp_project = "<gcp-project-id>"
   # gcp_location is the bucket location that spans the regions you're

>    # going to schedule across in Batch.  If you are running on one

Reviewer: > If you are running on one region, it can just be that region.

Unfortunately that doesn't quite work in the Australian setting. `location = "australia-southeast1"` and `storage_class = "MULTI_REGIONAL"` yields:
> Error: googleapi: Error 400: The combination of locationConstraint and storageClass you provided is not supported for your project, invalid

Unfortunately there's only a US, EU, and ASIA setting for multi-region buckets, so that unfortunately would be pretty expensive from Australia.

Author: OK, I renamed gcp_location => batch_logs_bucket_location and added a batch_logs_bucket_storage_class tfvar for this.

   # region, it can just be that region.
   gcp_location = "<gcp-location>"
   gcp_region = "<gcp-region>"
   gcp_zone = "<gcp-zone>"
   domain = "<domain>"
   ```

- Run `terraform init`.

- Run `terraform apply -var-file="$HOME/.hail/global.tfvars"`.
  Terraform has created a GKE cluster named `vdc`.  Configure
  `kubectl` to point at the vdc cluster:

  ```
  gcloud container clusters get-credentials --zone <gcp-zone> vdc
  ```

> - Go to the Google Cloud console, VPC networks > internal > Private

Reviewer: ```suggestion
- Go to the Google Cloud console, VPC networks > default > Private
```

- Create a VM on the internal network, 100GB, Ubuntu 20.04 TLS, allow
  full access to all Cloud APIs.  10GB will run out of space.  We
  assume the rest of the commands are run on the VM.

- Standardize file permissions.  This is for docker, which considers
  permissions for caching.  Run `echo 'umask 022' > ~/.profile`.  You
  will need to log out/in or run `umask 022`.

- Clone the Hail Github repository:

  ```
  git clone https://github.com/hail-is/hail.git
  ```

- Install some dependencies on the VM:

  ```
  sudo apt update
  sudo apt install -y docker.io python3-pip openjdk-8-jre-headless jq
  sudo snap install --classic kubectl
  sudo usermod -a -G docker $USER
  gcloud -q auth configure-docker
  gcloud container clusters get-credentials --zone <gcp-zone> vdc
  python3 -m pip install -r $HOME/hail/docker/requirements.txt
  ```

  You will have to log out/in for the usermod to take effect.

- Update $HAIL/config.mk with your infrastructure settings.  You can
  get settings from the default/global-config secret:

  ```
  kubectl -n default get secret global-config -o json | jq '.data | map_values(@base64d)'
  ```

- In `$HAIL/docker/third-party` run:

  ```

>   PROJECT=<gc-project-id> ./copy_images.sh

Reviewer: ```suggestion
  PROJECT=<gcp-project-id> ./copy_images.sh
```

  ```

  Wait for the `build-batch-worker-image` instance to be stopped.  Then run:

  ```
  make create-worker-image
  ```

- Bootstrap the cluster by running:

  ```
  HAIL_CI_UTILS_IMAGE=gcr.io/<gcp-project>/ci-utils:latest \
    HAIL_CI_BUCKET_NAME=dummy \
    KUBERNETES_SERVER_URL='<k8s-server-url>' \
    HAIL_DEFAULT_NAMESPACE='default' \
    HAIL_DOMAIN=<domain> \
    HAIL_GCP_ZONE=<gcp-zone> \
    HAIL_GCP_PROJECT=<gcp-project> \
    PYTHONPATH=$HOME/hail/ci:$HOME/hail/batch:$HOME/hail/hail/python \
    python3 $HAIL/ci/bootstrap.py hail-is/hail:main $(git rev-parse HEAD) test_batch_0
  ```

- Create the initial (developer) user:

  ```
  HAIL_CI_UTILS_IMAGE=gcr.io/<gcp-project>/ci-utils:latest \
    HAIL_CI_BUCKET_NAME=dummy \
    KUBERNETES_SERVER_URL='<k8s-server-url>' \
    HAIL_DEFAULT_NAMESPACE='default' \
    HAIL_DOMAIN=<domain> \
    HAIL_GCP_ZONE=<gcp-zone> \
    HAIL_GCP_PROJECT=<gcp-project> \
    PYTHONPATH=$HOME/hail/ci:$HOME/hail/batch:$HOME/hail/hail/python \
    python3 $HAIL/ci/bootstrap.py --extra-code-config '{"username":"<username>","email":"<email>"}' hail-is/hail:main $(git rev-parse HEAD) create_initial_user
  ```

>   Additional users can be added by the intial user by going to auth.hail.is/users.

Reviewer: ```suggestion
  Additional users can be added by the initial user by going to auth.<domain>/users.
```

This is a work in progress to use Terraform to manage our cloud
infrastructure.

Instructions:

> - You will need a GCP project.  Create a service account for Terraform

Reviewer: Might be worth mentioning the required APIs that should be enabled, otherwise it'll take a few Terraform runs. E.g. https://console.developers.google.com/marketplace/details/google-cloud-platform/compute-engine, https://console.developers.google.com/apis/library/cloudresourcemanager.googleapis.com.

Author: Did you have to enable servicenetworking?  I had a resource for that:

```
resource "google_project_service" "service_networking" {
  service = "servicenetworking.googleapis.com"
}
```

and I'm wondering if it didn't work.

Reviewer: Yes, here's the error message I got:

```
google_project_service.service_networking: Creating...
google_compute_global_address.gateway: Creation complete after 15s [id=projects/<redacted>/global/addresses/gateway]
kubernetes_secret.global_config: Modifying... [id=default/global-config]
kubernetes_secret.global_config: Modifications complete after 1s [id=default/global-config]
google_project_service.service_networking: Still creating... [10s elapsed]
google_project_service.service_networking: Still creating... [20s elapsed]
google_project_service.service_networking: Creation complete after 28s [id=<redacted>/servicenetworking.googleapis.com]

Error: googleapi: Error 403: Service Networking API has not been used in project <redacted> before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/servicenetworking.googleapis.com/overview?project=<redacted> then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry., accessNotConfigured
```

After enabling the API, I still got a 403, but that was because of missing permissions for the service account:

```
google_service_networking_connection.private_vpc_connection: Creating...

Error: googleapi: Error 403: The caller does not have permission, forbidden
```

Author: Ah, looks like it needs an explicit dependency.  That's annoying and race condition-y.  I'll just put the gcloud enable in the README.  Thanks!

Reviewer: https://github.com/terraform-google-modules/terraform-google-project-factory/tree/master/modules/project_services#project-api-activation might be worth a try.

Author: I looked into this.  It just expands into a set of "google_project_service" resources like I had originally, but doesn't appear to address the dependency issue.

>   with Editor role, and create a service account key and place it in

Reviewer: The service account will also need the Service Networking Admin role, otherwise the google_service_networking_connection.private_vpc_connection creation will fail.

Author: I fixed this by making the terraform service account Owner instead of Editor.
